plugins/dell-dock/fu-dell-dock-usb-usb4.c

```diff
@@ -0,0 +1,654 @@
+
+		buf += nbytes;
+		nvm_addr += nbytes;
+		length -= nbytes;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_dell_dock_usb4_hub_nvm_write (FuDevice *device, const guint8 *buf,
+				 guint32 length, guint32 nvm_addr,
+				 GError **error)
+{
+	guint8 metadata[4];
+	guint8 op_status = 0;
+	guint32 bytes_done = 0;
+	guint32 bytes_total = length;
+
+	if (nvm_addr % 4) {
+		g_prefix_error (error,
+			    "Invalid NVM write offset 0x%x, must be DW aligned",
+			    nvm_addr);
+		return FALSE;
+	}
+	if (length < 64 || length % 64) {
+		g_prefix_error (error,
+			     "Invalid NVM length 0x%x, must be 64 byte aligned",
+			     length);
+		return FALSE;
+	}
+
+	/* 1. Set initial offset, must be DW aligned */
+	fu_dell_dock_usb4_write_u32to_buf (NVM_OFFSET_TO_METADATA(nvm_addr),
+					   metadata);
+
+	if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_SET_OFFSET,
+					      metadata, &op_status,
+					      error)) {
+		g_prefix_error (error,
+				"hub firmware NVM set offset error, status %d",
+				op_status);
+		return FALSE;
+	}
+
+	/* 2 Write data in 64 byte blocks */
+	fu_device_set_progress_full (device, bytes_done, bytes_total);
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_WRITE);
+	while (length > 0)  {
+		/* write data to mbox data regs */
+		if (!fu_dell_dock_usb4_mbox_data_write (device, buf, 64,
+							error)) {
+			g_prefix_error (error,
+					"hub firmware mbox data write error");
+			return FALSE;
+		}
+		/* ask hub to write 64 bytes from data regs to NVM */
+		if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_WRITE,
+						      NULL, &op_status, error)) {
+			g_prefix_error (error,
+			    "hub firmware NVM write operation error, status %d",
+			    op_status);
+			return FALSE;
+		}
+		buf += 64;
+		length -= 64;
+		fu_device_set_progress_full (device, bytes_done += 64,
+					     bytes_total);
+	}
+	fu_device_set_status (device, FWUPD_STATUS_DEVICE_BUSY);
+	return TRUE;
+}
+
+/* Issue a nvm auth write operation */
+static gboolean
+fu_dell_dock_usb4_hub_nvm_authenticate (FuDevice *device, GError **error)
+{
+	guint8 op_status;
+
+	if (!fu_dell_dock_usb4_hub_operation (device, OP_NVM_AUTH_WRITE,
+					       NULL, &op_status, error)) {
+		g_prefix_error (error,
+				"Note: Succesful NVM Auth operation may cause "
+				"immediate GR reset? and thus problems to read "
+				"NVM Auth operation status. ");
+		return FALSE;
+	}
+	switch (op_status) {
+		case 0:
+			g_clear_error (error);
+			return TRUE;
+			break;
+		case 1:
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "NVM Auth write operation error");
+			break;
+		case 2:
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "NVM Auth operation status: Retry NVM write");
+			break;
+		default:
+			g_set_error (error,
+				     G_IO_ERROR,
+				     G_IO_ERROR_FAILED,
+				     "NVM Auth operation status unknown");
+			break;
+	}
+	/* FIXME pass op_status to caller in some way */
+	return FALSE;
```

Comment:
I still think you should bring these cases into `fu_dell_dock_usb4_hub_operation` and drop `op_status`, it will clean up the error handling more nicely.